### ISSUE TITLE
Regression: page up/down on Kindle calls GotoPosRel instead of GotoViewRel

### ISSUE THREAD

**Issue by cccslater** [koreader/koreader#12380]
Hi,
I've been running into a problem using continuous view PDFs and zoom.

I have manual zoom enabled and set to 1x, so it basically works like zoom to width. The only reason I'm using manual is so I can set 0 overlap. This works fine in page mode

However, I can't use manual zoom with continuous mode due to #10339. That issue seems to have stalled, so the other option is adding variable overlap to the other zoom modes that do work with continuous, then I can just use zoom to width.

Is this possible?

Hope this makes sense! Thanks for the ridiculously fabulous software.

**Comment by jonnyl2** [comment]
I don't think your question is clear enough. Could you perhaps rephrase it or be more specific about what you'd like to happen/do?

**Comment by cccslater** [comment]
Yes, sorry. Thought it might be confusing.

I want to be able to set minimal/custom overlap in any zoom mode. Currently
it seems to me I can only do it in manual mode.

I would prefer to be using continuous rather than page mode, but I cannot
because I am in manual mode so I can set overlap (see the referenced issue).

To fix my specific problem, either manual zoom could be fixed for
continuous view (seems stalled); or custom overlap added for other zoom
modes.

Hope this is clearer 🙂

**Comment by jonnyl2** [comment]
It seems to me that you would like to have zoom-to-content-width, continuous mode, but no overlap? (And that it stays after reopening the book, because the 1.0 zoom mode apparently always switches back to page mode?)

You can remove the overlap by going to the Advanced settings and setting `DOVERLAPPIXELS` TO `0`. However, I recommend something a bit higher, e.g. 5, or some lines will be cut through right in the middle and not fully viewable on either screen page.)

(First open File browser, then top menu, Tools icon, More tools >, Advanced settings.)

**Comment by cccslater** [comment]
Thanks. I will give that a go.
Just out of interest, do you know why overlap is only shown in the GUI for
manual mode?
Is it just because the automatic modes are not meant to have many options
because otherwise they wouldn’t be quite so automatic and user friendly?

**Comment by cccslater** [comment]
Ok, I set DOVERLAP to 5. This works for zoom-to-width, but it seems each time I select ZTW after using a different zoom mode it goes back to full page zoom rather than width.

After I go on ZTW (with custom 5 overlap) then go on continuous mode, it is still zoomed to width but also has the same large overlap, despite my changing the default setting.

😞

**Comment by jonnyl2** [comment]
>Just out of interest, do you know why overlap is only shown in the GUI for
manual mode?

You're talking about the "Page overlap" setting in the top menu, right? I'm not sure to be honest.

> it seems each time I select ZTW after using a different zoom mode it goes back to full page zoom rather than width.

>After I go on ZTW (with custom 5 overlap) then go on continuous mode, it is still zoomed to width but also has the same large overlap, despite my changing the default setting.

Could you share screenshots of the overlap and your settings (the settings from the bottom menu's 2nd, 3rd, and 4th tabs, or create a new profile with current settings, select to show it as QuickMenu and share a screenshot of that)?

**Comment by NiLuJe** [comment]
Usual reminder that ReaderZooming is a nightmare so there's quite possibly subtle or not so subtle bug with the overlap handling to begin with, so all you good intentions may be for naught @jonnyl2 ;).

(But they're certainly appreciated ;)).

**Comment by cccslater** [comment]
Ok here is some multimedia:

- This is what I'm aiming for -- absolute minimum overlap per each 'scroll' down. This is on paged mode, I'd like to be able to do this on continuous.
https://github.com/user-attachments/assets/68c772cc-0711-4889-b2b8-d704dcd4a059

- These are my settings for minimal overlap on paged mode. Note that if I keep these settings but switch to continuous, the UI still says it's in manual zoom (and I now realise that it **is** -- it retains any manual zoom factor I set), but ignores the overlap setting.
![IMG_0590](https://github.com/user-attachments/assets/56bb1f18-8be2-4f09-a7a2-3b847cefa031)
![IMG_0591](https://github.com/user-attachments/assets/02bc0b44-d549-436f-927e-760dc94563ce)

- My next question was: why does overlap not appear as an option for zoom-to-width? You will see it below top, in manual mode; but not below bottom, in width mode.
![IMG_0592](https://github.com/user-attachments/assets/d2c74dd1-e17b-4416-a229-8d3ba6f2992e)
![IMG_0593](https://github.com/user-attachments/assets/db923796-8f7a-4bac-85db-8d29edd568bc)

- This is how zoom-to-width behaves **in page mode** when DOVERLAP = 0. It's closest to what I want, and I had hoped that it would work in continuous mode.
https://github.com/user-attachments/assets/ccedee1f-b45d-4465-8796-cfaeb87049b0

- However, this is what happens with DOVERLAP = 0 + zoom-to-width + continuous:
https://github.com/user-attachments/assets/a4a8096d-c86c-4d5b-9bff-066772d3d090

Basically it seems continuous view just completely ignores overlap settings. Initially I thought this was because manual zoom is known to work poorly with continuous view, but it seems a more general issue.

The other thing I spotted when DOVERLAP = 0 is if I am in zoom-to-width, then change to any other zoom mode, then go back on zoom-to-content, I am no longer in zoom-to-width but zoom-to-full. Usually it would remember to do zoom-to-width.

This is a lot of stuff. Sorry.

**Comment by jonnyl2** [comment]
I'm not sure what's going on in your last video. I would also expect it to page down with no overlap according to your description and settings. 

Let me share the settings I would use to achieve what you want. Perhaps it'll help you too:

1. First rotate your device to your desired orientation and adjust the rotation setting accordingly
2. Bottom menu, 2nd tab: Page Crop: `auto`. Margin: lowest setting
3. 3rd tab: Zoom to content (2nd icon), then Fit `width`
4. 4th tab: View Mode `continuous`, Page Gap `none` or `small` (up to you)
5. (Optional) If you'd like wider L/R margins, go back to the 2nd tab and increase the margins) 

Now, with DOVERLAPPIXELS set to 0 (or 5) I don't have to worry about any overlap settings in the UI anymore; it will never overlap more than the number of pixels selected until I reach the end of the book.

**Comment by cccslater** [comment]
Your instructions are exactly what I’ve been doing and it doesn’t work, see
the last video. Hence my feature request

**Comment by Commodore64user** [comment]
In case you don't know @cccslater, you can capture screenshots by pressing <kbd>ScreenKB</kbd> + <kbd>Menu</kbd>. then connect to a computer and look for the Screenshots folder.

**Comment by jonnyl2** [comment]
>Your instructions are exactly what I’ve been doing and it doesn’t work, see
the last video. Hence my feature request

Then I'm not sure what's going on there. On my Kobo I've never had a problem with these exact settings. Looks more like a bug from what I can tell. (Does Kindle perchance have some extra quirks with ReaderZooming, @NiLuJe?)

**Comment by NiLuJe** [comment]
It shouldn't, the only thing I could think of would be some *inaccessible* settings because of the lack of touch panel and the bottom panel settings, but that should have been obvious ;).

**Comment by cccslater** [comment]
I have been poking around in the source — I see that in `readerpaging.lua` there are
- `overlap = Screen:scaleBySize(G_defaults:readSetting("DOVERLAPPIXELS")),`
- `local offset = Geom:new{
            x = 0,
            y = last_visible_area.h - overlap
        }`
- `local offset = Geom:new{
            x = 0,
            y = -first_page_state.visible_area.h + overlap
        }`

Am I correct that the paging in continuous mode is handled in `readerview.lua`? It seems that way when I have added a debug statement (in `drawScrollPages`). 
If this is the right file, there is no reference to `DOVERLAP` anywhere which would explain why continuous mode ignores my overlap setting.

Am I looking in the right direction?

**Comment by NiLuJe** [comment]
This mught be spread across readerzooming, readerpaging & readerview. Good luck (you'll need it).

**Comment by cccslater** [comment]
Ok, in this section of `readerpaging.lua`:
```
function ReaderPaging:onGotoPosRel(diff)
    if self.view.page_scroll then
        self:onPanningRel(100*diff)
    else
        self:onGotoPageRel(diff)
    end
    self:setPagePosition(self:getTopPage(), self:getTopPosition())
    return true
end
```

If I change `100*diff` to something larger then I get nice big jumps. On landscape on my K4, 550 looks pretty good, which makes me wonder if I can change `100*diff` to something like `(height - DOVERLAPPIXELS)*diff` to solve my problem. I think I should be able to work out this bit.

What I'm now finding is that most of the time I get a nice big jump with minimal overlap — however, if I am near the top of the next page, it does a small jump to the top of that page, not a consistent big jump.
You can see below the difference between 3rd and 4th 'next page' press is too small.
![1](https://github.com/user-attachments/assets/47495df9-3c86-46e1-825d-06913dda1e3c)

I would LOVE some help to work out what is going on here :-)

**Comment by NiLuJe** [comment]
I've never had the courage to figure out how this mess actually works, FWIW ;).

**Comment by cccslater** [comment]
Ok that’s interesting. Is it stuff carried over from a different project? The code is barely commented so it’s difficult to work out what it does 😭

**Comment by Frenzie** [comment]
He's being a bit dramatic. ;-P But are you showing a bug or the result of several hacks not quite working out?

**Comment by cccslater** [comment]
My goal is to have continuous mode listen to the DOVERLAPPIXELS setting.
What I'm showing in the gif is the result of my changing 100 to 550. Most of the time it does a nice big 550px jump, apart from near the top of a page where it just jumps to the top.

**Comment by Frenzie** [comment]
> My goal is to have continuous mode listen to the DOVERLAPPIXELS setting.

It already does. Which steps are required to make it not do so?

**Comment by cccslater** [comment]
I have never had it follow my DOVERLAPPIXELS with any settings. Below are some settings where it doesn't use my DOVERLAPPIXELS (currently = 5). You can see 'vertical overlap' is at 36, not 5. I can't change this in the UI because there is no overlap setting for zoom to width.
![Reader_2024-09-09_122355](https://github.com/user-attachments/assets/55257fd2-aa57-4ea2-9c25-f3928737a5c6)
![Reader_2024-09-09_122403](https://github.com/user-attachments/assets/9daac62f-239f-41f5-b2d3-6593fa6c273a)

**Comment by Frenzie** [comment]
30 (default):

<img src=https://github.com/user-attachments/assets/c6f2e407-9351-4c0f-a385-5aa6279b0d47 width=40%> <img src=https://github.com/user-attachments/assets/f05f46c1-075d-448a-9087-10940ec646cf width=40%>

300:
<img src=https://github.com/user-attachments/assets/c6f2e407-9351-4c0f-a385-5aa6279b0d47 width=40%> <img src=https://github.com/user-attachments/assets/48d7f431-e6c5-45ce-925e-3fff1028ec82 width=40%>

> I have never had it follow my DOVERLAPPIXELS with any settings. Below are some settings where it doesn't use my DOVERLAPPIXELS (currently = 5). You can see 'vertical overlap' is at 36, not 5. I can't change this in the UI because there is no overlap setting for zoom to width.

Those are two different things.

**Comment by Frenzie** [comment]
PS UI for that other setting here

<img src=https://github.com/user-attachments/assets/362747fe-5e54-49b1-b3ef-40ce2a7cd2c7 width=40%>

**Comment by cccslater** [comment]
>Those are two different things.

Do you mean 'there is no overlap setting for zoom to width' is a different thing? I accept that.

Here is what actually happens on my kindle with different DOVERLAP settings.

30:
![DOVERLAP30-1](https://github.com/user-attachments/assets/2d4e6ed0-71a3-42c0-83d7-e0c475962a78)
![DOVERLAP30-2](https://github.com/user-attachments/assets/8ff052fa-8cbb-4355-904f-9f7a5a09cc57)

5:
![DOVERLAP5-1](https://github.com/user-attachments/assets/0a73cb55-ac60-40f9-800a-43cd11672477)
![DOVERLAP5-2](https://github.com/user-attachments/assets/cecfdc80-c148-4df3-875f-01fbcf92f761)

They are exactly the same. I promise you they are different screenshots.

> PS UI for that other setting here

Ok. I wasn't sure where 36 had come from. Thanks.

**Comment by cccslater** [comment]
The ONLY places DOVERLAPPIXELS is used is `overlap = Screen:scaleBySize(G_defaults:readSetting("DOVERLAPPIXELS")),`, then later `overlap` is used in `ReaderPaging:onScrollPageRel`. This isn't called when I press next page in continuous mode, hence why I don't think it is listening to my setting.

**Comment by Frenzie** [comment]
> Do you mean 'there is no overlap setting for zoom to width' is a different thing? I accept that.

`DOVERLAPPIXELS` is `DOVERLAPPIXELS`, not `Vertical overlap`. Its effect in continuous mode is as shown here: https://github.com/koreader/koreader/issues/12380#issuecomment-2337880015.

> This isn't called when I press next page in continuous mode, hence why I don't think it is listening to my setting.

This is why the issue templates asks for exact steps to reproduce. ;-) On which device on which version of KOReader are you pressing which button?

**Comment by cccslater** [comment]
> DOVERLAPPIXELS is DOVERLAPPIXELS, not Vertical overlap. Its effect in continuous mode is as shown here: https://github.com/koreader/koreader/issues/12380#issuecomment-2337880015.

When you say 'its effect' -- do you mean DOVERLAPPIXELS effect or Vertical overlap effect?

> This is why the issue templates asks for exact steps to reproduce. ;-) On which device on which version of KOReader are you pressing which button?

I am on v2024.07.172-gadbf3be6e_2024-09-07. I am on a Kindle 5 NT D01100. I am pressing the next page button (RPgFwd), or any of the side navigation buttons. I didn't include this info initially because I thought I was doing a feature request not a bug report :0

**Comment by Frenzie** [comment]
> When you say 'its effect' -- do you mean DOVERLAPPIXELS effect or Vertical overlap effect?

DOVERLAPPIXELS.

**Comment by cccslater** [comment]
Very bizarre. To me the event chain looks like this:

`RPgFwd` calls `onGotoPosRel` calls `onPanningRel` calls `onScrollPanRel`

None of these functions use `overlap` anywhere. Is the emulator calling different functions?

**Comment by Frenzie** [comment]
No, RPgFwd is RPgFwd. (Which is to say, regular page down is mapped as RPgFwd.)

**Comment by NiLuJe** [comment]
Stupid question: *How* did you attempt to set `DOVERLAPPIXELS`?

**Comment by Frenzie** [comment]
For reference:

```lua
-- ./defaults.custom.lua
return {
    ["DOVERLAPPIXELS"] = 300,
}
```

**Comment by cccslater** [comment]
> Stupid question: _How_ did you attempt to set `DOVERLAPPIXELS`?

In advanced settings in the UI. It has definitely changed the setting because it is reflected in `defaults.custom.lua`.

I'm not an expert, but I think it's because onScrollPanRel is called which doesn't refer to `overlap` at all.

I have modified the function as follows:

```lua
function ReaderPaging:onScrollPanRel(diff)
    local overlap = self.overlap
    if diff == 0 then return true end
    logger.dbg("pan relative height:", diff)
    
    -- Remove this
    --local offset = Geom:new{x = 0, y = diff}
    
    local blank_area = Geom:new()
    blank_area:setSizeTo(self.view.visible_area)
    local new_page_states
    if diff > 0 then
    	
    	-- Add this
    	local offset = Geom:new{x = 0, y = self.view.visible_area.h - overlap}
        
        -- pan to scroll down
        local first_page_state = copyPageState(self.view.page_states[1])
        new_page_states = self:genPageStatesFromTop(
        first_page_state, blank_area, offset)
    elseif diff < 0 then

    	-- Add this
    	local offset = Geom:new{x = 0, y = overlap - self.view.visible_area.h}
    	
        local last_page_state = copyPageState(
            self.view.page_states[#self.view.page_states])
        new_page_states = self:genPageStatesFromBottom(
            last_page_state, blank_area, offset)
    end
    if #new_page_states == 0 then
        -- if we are already at the top of first page or bottom of the last
        -- page, new_page_states will be empty, in this case, nothing to update
        return true
    end
    self.view.page_states = new_page_states
    -- update current pageno to the very last part in current view
    self:_gotoPage(self.view.page_states[#self.view.page_states].page,
                   "scrolling")
    UIManager:setDirty(self.view.dialog, "partial")
    return true
end
```

Instead of just the line `local offset = Geom:new{x = 0, y = diff}` I now have separate lines for scrolling up and down, calculating with `overlap` instead of `diff`. This makes it much more similar to `onScrollPageRel`, which does use `overlap`.

My new function mostly works as expected except it still does small jumps when near the top of the page!

**Comment by cccslater** [comment]
No new function needed! If I remap the page buttons to call `GotoViewRel` instead of `GotoPosRel`, all my problems disappear. :-)

**Comment by Frenzie** [comment]
> I'm not an expert, but I think my issue is because onScrollPanRel is called which doesn't refer to overlap at all.

Did you hack something to make it call ScrollPanRel? The arrow keys (as in dpad) would call that but not page up/down.

**Comment by Frenzie** [comment]
Regression in #12174 then.

**Comment by Frenzie** [comment]
If you're motivated you could check that for example 2024.03 works as expected.

**Comment by cccslater** [comment]
I'm motivated as heck. How do I install an older version?

**Comment by Frenzie** [comment]
Rename the current folder to koreader_bck and toss the old version in place? ^_^

Unless Kindle requires something else besides that.

**Comment by Frenzie** [comment]
Yes, the original code clearly shows `GotoViewRel` for page up/page down and `GotoPosRel` for up/down.
https://github.com/Commodore64user/KOReader_fork/blame/36d2e3cf74a536fea096749809dbff989fd240e7/frontend/apps/reader/modules/readerpaging.lua#L55-L73

**Comment by cccslater** [comment]
Yup, the older version works as I'd expect :-)

**Comment by Frenzie** [comment]
Please feel free to submit a quick PR to adapt those four lines. ^_^

**Comment by cccslater** [comment]
Ok! It's been a long time since I've done a PR but I'll do my best.
Which 4 lines do you mean? I can only see 2, the last lines in https://github.com/koreader/koreader/issues/12380#issuecomment-2338095256

**Comment by Frenzie** [comment]
Indeed, yes. I mean like this for all four lines:

```diff
diff --git a/frontend/apps/reader/modules/readerpaging.lua b/frontend/apps/reader/modules/readerpaging.lua
index 68a86a623..3af599ba2 100644
--- a/frontend/apps/reader/modules/readerpaging.lua
+++ b/frontend/apps/reader/modules/readerpaging.lua
@@ -53,7 +53,7 @@ function ReaderPaging:onGesture() end
 function ReaderPaging:registerKeyEvents()
     if Device:hasDPad() and Device:useDPadAsActionKeys() then
         if G_reader_settings:isTrue("left_right_keys_turn_pages") then
-            self.key_events.GotoNextPos = { { { "RPgFwd", "LPgFwd", "Right", " " } }, event = "GotoPosRel", args = 1, }
+            self.key_events.GotoNextPage = { { { "RPgFwd", "LPgFwd", "Right", " " } }, event = "GotoViewRel", args = 1, }
             self.key_events.GotoPrevPos = { { { "RPgBack", "LPgBack", "Left" } }, event = "GotoPosRel", args = -1, }
         elseif G_reader_settings:nilOrFalse("left_right_keys_turn_pages") then
             self.key_events.GotoNextChapter = { { "Right" }, event = "GotoNextChapter", args = 1, }
```

**Comment by poire-z** [comment]
Pinging @Commodore64user (if this change would sabotage his feature, or unsabotage his sabotage).

**Comment by Commodore64user** [comment]
Well @poire-z, the reason we went with that one was because the "would be one used here" will not display the full content of a large PDF page when in paged mode. Say a pdf is divided in page 23 part a and page 23 part b (cause it doesn't fit). The one we went with displays upon each page turn: 23.a > 23.b > 24.a > 24.b

What is being suggested as a change does 23.a > 24.a > 25.a instead.

I don't really read PDFs so it won't affect me but, i don't (personally) believe it should be changed in master but patched instead.
